devito/ir/equations/equation.py

```diff
@@ -0,0 +1,66 @@
+from collections import OrderedDict
+
+import sympy
+
+from devito.ir.support import Interval, IterationSpace, Stencil
+from devito.symbolics import dimension_sort, indexify
+
+__all__ = ['Eq']
+
+
+class Eq(sympy.Eq):
+
+    """
+    A new SymPy equation with an associated iteration space.
+
+    All :class:`Function` objects within ``expr`` get indexified and thus turned
+    into objects of type :class:`types.Indexed`.
+
+    An iteration space is an object of type :class:`Space`. It represents the
+    data points accessed by the equation along each :class:`Dimension`. The
+    :class:`Dimension`s are extracted directly from the equation.
+    """
+
+    def __new__(cls, input_expr, subs=None):
+        # Sanity check
+        assert isinstance(input_expr, sympy.Eq)
+
+        # Indexification
+        expr = indexify(input_expr)
+
+        # Apply caller-provided substitution
+        if subs is not None:
+            expr = expr.xreplace(subs)
+
+        expr = super(Eq, cls).__new__(cls, expr.lhs, expr.rhs, evaluate=False)
+        expr.is_Increment = getattr(input_expr, 'is_Increment', False)
+
+        # Get the accessed data points
+        stencil = Stencil(expr)
+
+        # Well-defined dimension ordering
+        ordering = dimension_sort(expr, key=lambda i: not i.is_Time)
```

Comment:
Why does time get special treatment here? If we had a custom dimension `k` and a `TimeFunction` `f(k, t, x, y)`, would the implied ordering be override by this line?

Reply:
yes; this is what we have in ``master`` too. This sounds bad, I agree, but it allows us to handle a few corner cases in which the time loop ended up inside an outer dimension, while it was conceptually expected to be the outermost iteration dimension. I already spoke to @mloubout about this, and for the time being we agreed to leave it that way. We can definitely think of a better solution, but this would be non trivial (I tried, already). This PR already improves a very large number of things, so I guess that at least this "issue", which has been around for a while, is for another time.

Reply:
Ok, but we really need that as an explicit test case somewhere in `test_operator.py` or equivalent, since again this is in no way obvious.

Reply:
I'm always willing to add tests, but how would you write such a test? Note that this is kinda unrelated to this PR, since this is the same behavior as in master; however, I'm happy to add it here.

Reply:
I mean, if you have an equation like this:
```
a(f, x, y) = b(t, x, y)*a(f, x, y)
```
where ``f`` is a "misc" dimension (YASK jargon), what tells you to generate:
```
for t
  for f
    for x
      for y
```
rather than

```
for f
  for t
    for x
      for y
```
?
surely ``y`` comes after ``x`` and both come after ``t/f``, but there's no way to say/infer/... whether ``t`` should appear after or before ``f``.
You may say "well, I see ``f`` first, so it should go first"; but in general we want the time loop to be the outermost. I can also come up with more pathological scenarios...

Reply:
Ok, this is probably stuff for a separate issue, but your example makes no sense:
* The RHS shares `x, y` but then multiplies `t * f`, essentially creating an outer product/tensor.
* The LHS then only knows one of the two dimensions of this tensor, thus implying a reduction in `t` (following our own convention about implied reductions).
* As a corollary, having `t` as the second-level loop is not only correct, but sensible, as it's performing a reduction in `t`_for all_ `f`.

> in general we want the time loop to be the outermost.

In fact, I see no justification in this example as to why we even want `t` outermost, afaict. The semantics will definitely not be harmed by `f` being outermost.

Most likely this example is too artificial to express what you want. I'm also not trying to stop this PR, but basically, whether in`master` or not, this "special treatment" could looks like a dirty hack we want to get rid of, unless someone can provide some sound justification for it that is documented somewhere.

Reply:
> his "special treatment" could looks like a dirty hack
I have no objections to calling this a hack; in fact, I probably agree with you. 

You can make the reduction explicit in my example if you prefer:
```
a(f, x, y) = a(f, x, y) + b(t, x, y)*a(f, x, y)
```
and now again you can detect ("read", rather than "detect", if the information is carried by the ``Eq``) that there's a reduction and set up a rule for dimension ordering in case of reductions, OR basically you're just saying the same exact thing that I said -- there is no unique way of scheduling the ``f`` and ``t`` loops because both orders are correct ! 

What I didn't say, as hidden in between the lines, is that preferring the time dimension is outermost is probably, *performance-wise*, the easy thing to do in such a case

Reply:
>  there is no unique way of scheduling the f and t loops because both orders are correct !

Yes, from the pure API semantics that's exactly right. So how do we decide that `t` is indeed special enough to be outermost? Well, if it's a `SteppingDimension` it would make sense to say "be as outermost as semantically permitted", as processing a single buffer-iteration is probably fastest. Maybe solving this becomes as easy as switching `is_Time` -> `is_Stepping`, and leaving a detailed comment (or as I said, test)?

In any case, advanced nitpicking, so ...